sqlglot/optimizer/qualify_columns.py

```diff
@@ -413,19 +488,46 @@ def _expand_stars(
             if not pivot_output_columns:
                 pivot_output_columns = [c.alias_or_name for c in pivot.expressions]
 
+    if dialect == "bigquery" and any(isinstance(col, exp.Dot) for col in scope.stars):
+        # Found struct expansion, annotate scope ahead of time
+        annotator.annotate_scope(scope)
+
     for expression in scope.expression.selects:
+        tables = []
+        structs = []
         if isinstance(expression, exp.Star):
             tables = list(scope.selected_sources)
             _add_except_columns(expression, tables, except_columns)
             _add_replace_columns(expression, tables, replace_columns)
-        elif expression.is_star and not isinstance(expression, exp.Dot):
-            tables = [expression.table]
-            _add_except_columns(expression.this, tables, except_columns)
-            _add_replace_columns(expression.this, tables, replace_columns)
+        elif expression.is_star:
+            if not isinstance(expression, exp.Dot):
+                tables = [expression.table]
+                _add_except_columns(expression.this, tables, except_columns)
+                _add_replace_columns(expression.this, tables, replace_columns)
+            elif dialect == "bigquery":
+                col = expression.find(exp.Column)
+                table_name = col.table if col else None
+                starting_struct = str(expression.this)
+                structs.append((col, starting_struct))
         else:
             new_selections.append(expression)
             continue
 
+        if structs:
+            struct_fields = _expand_struct_stars(scope, structs) or []
+
+            for struct_name in struct_fields:
```

Comment:
I think we can simplify this section a bit, if `structs` is empty we'll always get back an empty list from the expand helper:
```suggestion
        for struct_name in _expand_struct_stars(scope, structs):
```